### ISSUE TITLE
-Warray-bounds in net/phonet/pep.c

### ISSUE THREAD

**Issue by tpimh** [ClangBuiltLinux/linux#115]
```
net/phonet/pep.c:224:16: warning: array index 1 is past the end of the array (which contains 1 element) [-Warray-bounds]
        ph->data[0] = oph->data[1]; /* CTRL id */
                      ^         ~
./include/net/phonet/pep.h:66:3: note: array 'data' declared here
                u8              data[1];
                ^
net/phonet/pep.c:281:10: warning: array index 1 is past the end of the array (which contains 1 element) [-Warray-bounds]
        switch (hdr->data[1]) {
                ^         ~
./include/net/phonet/pep.h:66:3: note: array 'data' declared here
                u8              data[1];
                ^
net/phonet/pep.c:285:12: warning: array index 4 is past the end of the array (which contains 1 element) [-Warray-bounds]
                        switch (hdr->data[4]) {
                                ^         ~
./include/net/phonet/pep.h:66:3: note: array 'data' declared here
                u8              data[1];
                ^
net/phonet/pep.c:295:8: warning: array index 4 is past the end of the array (which contains 1 element) [-Warray-bounds]
                        if (hdr->data[4] == PEP_IND_READY)
                            ^         ~
./include/net/phonet/pep.h:66:3: note: array 'data' declared here
                u8              data[1];
                ^
net/phonet/pep.c:304:21: warning: array index 4 is past the end of the array (which contains 1 element) [-Warray-bounds]
                atomic_add(wake = hdr->data[4], &pn->tx_credits);
                                  ^         ~
./include/net/phonet/pep.h:66:3: note: array 'data' declared here
                u8              data[1];
                ^
net/phonet/pep.c:509:9: warning: array index 4 is past the end of the array (which contains 1 element) [-Warray-bounds]
        n_sb = hdr->data[4];
               ^         ~
./include/net/phonet/pep.h:66:3: note: array 'data' declared here
                u8              data[1];
                ^
net/phonet/pep.c:817:14: warning: array index 1 is past the end of the array (which contains 1 element) [-Warray-bounds]
        peer_type = hdr->other_pep_type << 8;
                    ^    ~~~~~~~~~~~~~~
./include/net/phonet/pep.h:66:3: note: array 'data' declared here
                u8              data[1];
                ^
net/phonet/pep.c:820:9: warning: array index 4 is past the end of the array (which contains 1 element) [-Warray-bounds]
        n_sb = hdr->data[4];
               ^         ~
./include/net/phonet/pep.h:66:3: note: array 'data' declared here
                u8              data[1];
                ^
8 warnings generated.
```

**Comment by nickdesaulniers** [comment]
Worrisome, but I think it's still safe since the headers come from `pnp_hdr()`.  I think that `data` should be a `u8*`, not `u8[1]`.

```diff
-                u8              data[1];
+                u8              *data;
```

**Comment by nathanchance** [comment]
Pending patch: https://github.com/nathanchance/patches/blob/master/linux/pending/0002-phonet-Fix-array-bounds-warnings.patch

I'm sure my justification could use some better wording but it clears up the warning just fine and GCC doesn't complain.

**Comment by nathanchance** [comment]
I went ahead and sent an email to the maintainers to see if they have any ideas since I'm not confident in my wording of Nick's suggestion: https://lore.kernel.org/lkml/20190108025420.GA26093@flashbox/

**Comment by nickdesaulniers** [comment]
Maybe `git blame <file> -L <lineno>` for the line in question can point us to a commit of the author, who we can then cc on the existing thread, asking for advice?

**Comment by nathanchance** [comment]
Unfortunately, that line was introduced by the maintainer in 9641458d3ec42def729fde64669abf07f3220cd5, who I directed the email to.

However, I went ahead and bumped the thread with a different email address for the maintainer.

**Comment by gwelymernans** [comment]
I have a suspicion that they meant to use `u8 data[0];` here. It looks like it's sending / receiving multi-byte messages as a payload (at least if I'm reading the code correctly). Does `[0]` work for you?

**Comment by nathanchance** [comment]
@gwelymernans yes, it does

```diff
diff --git a/include/net/phonet/pep.h b/include/net/phonet/pep.h
index b669fe6dbc3b..5bc9aed896fc 100644
--- a/include/net/phonet/pep.h
+++ b/include/net/phonet/pep.h
@@ -63,7 +63,7 @@ struct pnpipehdr {
 		u8		state_after_reset;	/* reset request */
 		u8		error_code;		/* any response */
 		u8		pep_type;		/* status indication */
-		u8		data[1];
+		u8		data[0];
 	};
 };
 #define other_pep_type		data[1]
```

On top of `v5.0-rc4`:

```
make -j16 ARCH=arm64 CC=clang HOSTCC=clang CROSS_COMPILE=aarch64-linux-gnu- allyesconfg modules_prepare
make -j16 ARCH=arm64 CC=clang HOSTCC=clang CROSS_COMPILE=aarch64-linux-gnu- M=net/phonet
```

**Comment by nathanchance** [comment]
Arnd's patch was accepted and queued in net, should make it in to 5.0 final: https://git.kernel.org/pub/scm/linux/kernel/git/davem/net.git/commit/?id=6321aa197547da397753757bd84c6ce64b3e3d89

**Comment by nickdesaulniers** [comment]
Thanks @arndb ! Please CC us or include a link to our issue tracker when you can. ⛷

**Comment by nathanchance** [comment]
He did CC me, I just forgot to add you or put a link to the posting :(

For the record: https://lore.kernel.org/lkml/20190219215359.391543-1-arnd@arndb.de/

**Comment by nathanchance** [comment]
Merged into mainline (it made it into 5.0): https://git.kernel.org/torvalds/c/6321aa197547da397753757bd84c6ce64b3e3d89
